[Wong-Jia-Jun] iP

Wong-Jia-Jun opened the pull request:
### DukeBot
___
DukeBot helps you remember things effortlessly. It's,

- text-based

- Super easy to pickup

- very ACCESSIBLE

Its commands are also **intuitive** : `todo` `delete`

Features:
- [ ] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (coming soon)


If you Java programmer, you can have a look at the SUPER readable code. :grinning: 
Here's an example method:
```java
private static void addTodo(String description, TaskList list) {
        ToDo tdItem = new ToDo(description);
        list.addTask(tdItem);
        System.out.println("\tGot it. I have added this task:\n" + "\t" + tdItem.toString());
        System.out.printf("\tNow you have %d tasks in the list.\n", list.size());
    }
```
All you need to do is,

1. download it from [here](https://github.com/Wong-Jia-Jun/ip).
2. double-click it.
3. add your tasks.
4. **_ENJOY_** the convenience it brings

> “The advance of technology is based on making it fit in so that you don't really even notice it, so it's part of everyday life.”
Bill Gates, Co-founder of Microsoft.

Commit:
Add Gradle support

Wong-Jia-Jun committed:
Update Duke.java

Wong-Jia-Jun committed:
Update Duke.java

Wong-Jia-Jun committed:
added skeleton for Tasks with level 3

Wong-Jia-Jun committed:
level 4 plus exception skele

Wong-Jia-Jun committed:
handle few errors/level 5

Wong-Jia-Jun committed:
level-6

Wong-Jia-Jun committed:
Update UI testing

Wong-Jia-Jun committed:
Add Save function

Wong-Jia-Jun committed:
Add date/Time for Deadline and Events tasks

Wong-Jia-Jun committed:
Resolve merge conflicts

Wong-Jia-Jun committed:
Resolve file reading error

Wong-Jia-Jun committed:
Merge branch 'master' into branch-Level-8

Wong-Jia-Jun committed:
Merge pull request #2 from Wong-Jia-Jun/branch-Level-8

Branch level 8

Wong-Jia-Jun committed:
Level-8

Wong-Jia-Jun committed:
Add MoreOOP stuff

Wong-Jia-Jun committed:
Merge branch 'master' of https://github.com/Wong-Jia-Jun/ip

Wong-Jia-Jun committed:
Allocate classes to different Packages

Wong-Jia-Jun committed:
Fix some commit issues

Wong-Jia-Jun committed:
add javadoc for Duke class

Wong-Jia-Jun committed:
Add Java docs

Wong-Jia-Jun committed:
Merge branch 'master' into A-JavaDoc

Wong-Jia-Jun committed:
Merge pull request #3 from Wong-Jia-Jun/A-JavaDoc

A java doc

Wong-Jia-Jun committed:
Merge pull request #4 from Wong-Jia-Jun/add-gradle-support

Add Gradle support

Wong-Jia-Jun committed:
Change name of package

Wong-Jia-Jun committed:
Merge branch 'master' of https://github.com/Wong-Jia-Jun/ip

Wong-Jia-Jun committed:
Add JUnit

Wong-Jia-Jun committed:
Add find function

Wong-Jia-Jun committed:
Merge pull request #5 from Wong-Jia-Jun/branch-Level-9

Add find function

Dangabit commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,196 @@
+     * @param list Tasklist  which actions are to be performed on
+     * @return int of -1 if bye instruct is parsed else 1 is return to signal program to continue running
+     */
+    public static int parse(String instct, TaskList list) throws  IOException  {
```
This method seems to be quite long. Perhaps break it up into smaller methods (such as a method for each command, or a new class for each command) would allow for easier code maintenance in the future?

Dangabit commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+     * @return ArrayList<Task>
+     */
+
+    public  ArrayList<Task> readnWriteData(){
```
Similar issue of a long method here.

Dangabit commented on src/main/java/Duke/Ui.java:
```diff
@@ -0,0 +1,107 @@
+    /**
+     * The method of Underline() to print out a separating line
+     */
+    public static void Underline(){
```
To follow coding standard.
```suggestion
    public static void underline()
```

Dangabit commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+package Duke;
+
+import Duke.Tasks.*;
```
Wildcard imports violate coding standards.

Dangabit commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,196 @@
+            }
+            else if (instct.split(" ")[0].equals("bye")) {
+                Ui.sayBye();
+                return -1;
```
There may be an issue of magic numbers here. I understand that this is used for the flag. Perhaps a change to boolean values or more explicit indication of the meaning behind the numbers (through variable declaration) may be better?

Dangabit commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,196 @@
+    public static int parse(String instct, TaskList list) throws  IOException  {
+        String cmd = instct.split(" ")[0];
+        try {
+            if (instct.split(" ")[0].equals("list") ) {
```
I think using Java enums will be good here to abstract the concept of commands.

Dangabit reviewed (commented):
Code quality wise there can be some improvements!
Some suggestions have been included for both code quality and standards.

Wong-Jia-Jun committed:
Update javaDoc

Wong-Jia-Jun committed:
Tweak or Code standards

Wong-Jia-Jun committed:
Merge pull request #6 from Wong-Jia-Jun/A-CodingStandard

Tweak or Code standards

Wong-Jia-Jun committed:
Add GUI support in progress

Wong-Jia-Jun committed:
Add gui support

Wong-Jia-Jun committed:
Add branch

forgot to add

Wong-Jia-Jun committed:
Merge pull request #7 from Wong-Jia-Jun/branch-Level-10

Add branch

Wong-Jia-Jun committed:
Add assertions

Commands that add task to TaskList: add assertions to ensure that new tasks is added to the list

No current measures to detect an error when adding tasks.

An extra layer of protection  to prompt User if an underlying problem is happening

Wong-Jia-Jun committed:
Merge pull request #8 from Wong-Jia-Jun/branch-A-Assertions

Add assertions

Wong-Jia-Jun committed:
Abstract out executions in parse method to improve code quality

Abstract out executions in parse method to improve code quality

parse method a little long and too cluttered, improve readability and  easy maintenance

pull out the formatting of String response to individual methods

Wong-Jia-Jun committed:
Merge pull request #9 from Wong-Jia-Jun/A-CodeQuality

Abstract out executions in parse method to improve code quality

Wong-Jia-Jun committed:
Update format and confusing variable name in TaskList class

static variable in Tasklist class called listOfTask is type ArrayList but in other classes same name variable is type TaskList

Change variable name in TaskList class to arrayListOfTasks to avoid confusion

Minor fixes, so as to not build further confusion and improve readability

Wong-Jia-Jun committed:
Add background effect to image

dawg420 commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,220 @@
+ * Contains Parser object that deals with making sense of the user command
+ */
+public class Parser {
+    private final static DateTimeFormatter  timeFormat = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    private final static DateTimeFormatter  HrFormat = DateTimeFormatter.ofPattern("HHmm");
```
constants should be in caps

dawg420 commented on src/main/java/Duke/Tasks/Task.java:
```diff
@@ -0,0 +1,80 @@
+        return (isDone ? "[X]" : "[ ]"); // mark done task with X
+    }
+
+    /**
+     * The getter method for getting description
+     *
+     */
```
Comments could perhaps follow JavaDocs format more closely. (Include @params and @return

dawg420 commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+import java.time.format.DateTimeFormatter;
+import java.time.LocalTime;
+
+/**
+ * Contains Storage object that deals with loading listOfTasks from the file and saving listOfTasks in the file
+ */
```
I like your clear and concise description of the class

dawg420 reviewed (commented):
Generally well formatted!

Wong-Jia-Jun committed:
Merge pull request #10 from Wong-Jia-Jun/A-CodeQuality

A code quality

Wong-Jia-Jun committed:
Add feature to detect duplicates

Add ability to detect and prevent duplicates as an exception

It is currently possible to add duplicate tasks which will flood the list with unnecessary information as some are duplicates or duplicates with more information

Unifying all those duplicated code improves the code quality.

Lets add a duplicateException which will handle duplicates by prompting user when performing any commands which adds new tasks to list

Doing so will remind user that task is already present so to either delete old one and add new updated one or just not perform the command since its already present

Wong-Jia-Jun committed:
Merge pull request #12 from Wong-Jia-Jun/branch-C-DetectDuplicates

Add feature to detect duplicates

Wong-Jia-Jun committed:
Fix violations according to soc-se-bot

Wong-Jia-Jun committed:
abstract few more lines in parse method

Wong-Jia-Jun committed:
Customize responses bot gives to various commands

Wong-Jia-Jun committed:
Update ReadMe.md

Wong-Jia-Jun committed:
Fix empty description exception

Wong-Jia-Jun committed:
Remove not needed line

Wong-Jia-Jun committed:
Fix parser for EmptyDescription Error

Wong-Jia-Jun committed:
Create Ui.png
